File: Data Structures & Algorithms/best-time-to-buy-and-sell-stock-ii/submission-0.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int currentProfit = 0;

        int boughtAt = 0;

        for (int i = 1; i < prices.length; i++) {
            if (prices[i] > prices[i - 1]) {
                currentProfit = Math.max(
                    prices[i] - prices[boughtAt], currentProfit);
            } else {
                // book the profit
                maxProfit += currentProfit;
                currentProfit = 0;
                boughtAt = i;
            }
        }
        if (currentProfit != 0) maxProfit += currentProfit;

        return maxProfit;
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        minTillNow = 2 ** 32 - 1
        maxProfitTillNow = 0

        for price in prices:
            minTillNow = min(minTillNow, price)

            maxProfitTillNow = max(
                maxProfitTillNow,
                price - minTillNow,
            )

        return maxProfitTillNow